File: src/main/java/com/mybar/bartender/service/UserService.java
package com.mybar.bartender.service;


import com.mybar.bartender.dto.RegistrationUserDto;
import com.mybar.bartender.dto.UserDto;
import com.mybar.bartender.model.User;
import com.mybar.bartender.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class UserService implements UserDetailsService {
    @Lazy
    @Autowired
    private PasswordEncoder passwordEncoder;
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private RoleService roleService;

    @Override
    @Transactional
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        User user = findByUsername(username).orElseThrow(() -> new UsernameNotFoundException(
                String.format("Пользователь '%s' не найден", username)
        ));
        return new org.springframework.security.core.userdetails.User(
                user.getName(),
                user.getPassword(),
                user.getRoles().stream().map(role -> new SimpleGrantedAuthority(role.getName())).collect(Collectors.toList())
        );
    }

    public Optional<User> findByUsername(String username) {
        return Optional.ofNullable(userRepository.findByName(username));
    }

    public Optional<User> findByEmail(String email) {
        return Optional.ofNullable(userRepository.findByEmail(email));
    }

    public List<UserDto> getAllUsers() {
        return ((List<User>) userRepository.findAll()).stream().map(x -> new UserDto(x.getId(), x.getName(), x.getEmail())).collect(Collectors.toList());
    }

    public User createNewUser(RegistrationUserDto registrationUserDto) {
        User user = new User();
        user.setName(registrationUserDto.getUsername());
        user.setEmail(registrationUserDto.getEmail());
        user.setPassword(passwordEncoder.encode(registrationUserDto.getPassword()));
        user.setRoles(List.of(roleService.getUserRole()));
        return userRepository.save(user);
    }
}

File: src/main/kotlin/com/mybar/bartender/service/UserService.kt
package com.mybar.bartender.service

import com.art.playersapi.exception.EntityNotFoundException
import com.mybar.bartender.dto.RegistrationUserDto
import com.mybar.bartender.dto.UserDto
import com.mybar.bartender.model.Role
import com.mybar.bartender.model.User
import com.mybar.bartender.repository.UserRepository
import org.springframework.context.annotation.Lazy
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class UserService(
    @Lazy
    private var passwordEncoder: PasswordEncoder,
    private val userRepository: UserRepository,
    private val roleService: RoleService,
) : UserDetailsService {
    override fun loadUserByUsername(username: String?): UserDetails {
        val user = findByUsername(username)
            ?: throw UsernameNotFoundException(String.format("Пользователь '%s' не найден", username))
        return org.springframework.security.core.userdetails.User(
            user.name,
            user.password,
            user.roles!!.stream().map { x ->
                SimpleGrantedAuthority(
                    x!!.name
                )
            }.collect(Collectors.toList())
        )
    }

    fun findByUsername(username: String?): User? {
        return userRepository.findByName(username)
    }

    fun findByEmail(email: String?): User? {
        return userRepository.findByEmail(email!!)
    }

    fun getAllUsers(): List<UserDto> {
        return (userRepository.findAll() as List<User>).stream().map { x: User ->
            UserDto(
                x.id!!,
                x.name,
                x.email
            )
        }.collect(Collectors.toList())
    }

    fun createNewUser(registrationUserDto: RegistrationUserDto): User {
        val user = User(
            name = registrationUserDto.username,
            email = registrationUserDto.email,
            password = passwordEncoder.encode(registrationUserDto.password),
            roles = java.util.List.of(roleService.getUserRole())
        )
        return userRepository.save(user)
    }
}